app/components/SignUpForm.tsx

```diff
@@ -72,38 +100,59 @@ export default function SignUpForm() {
           placeholder="Enter your email address"
           type="text"
           name="email"
-          onChange={handleChange}
-          value={formData.email}
+          register={register}
+          rules={signUpRules.email}
+          errors={errors}
+          value={watch("email")}
         />
         <Input
           icon={<User />}
           label="Username"
           placeholder="Enter your username"
           type="text"
           name="username"
-          onChange={handleChange}
-          value={formData.username}
+          register={register}
+          required={true}
+          rules={signUpRules.username}
+          errors={errors}
+          value={watch("username")}
         />
         <Input
           icon={<Lock />}
           label="Password"
           placeholder="Enter your password"
           type="password"
           name="password"
-          onChange={handleChange}
-          value={formData.password}
+          register={register}
+          required={true}
+          rules={signUpRules.password}
+          errors={errors}
+          value={watch("password")}
         />
         <Input
           icon={<Lock />}
           label="Confirm Password"
           placeholder="Confirm your password"
           type="password"
           name="confirmPassword"
-          onChange={handleChange}
-          value={formData.confirmPassword}
+          register={register}
+          required={true}
```

Comment:
I see you have repetead times `required={true}`
The required rule is already included in the rules object under the register function, and this handles the validation directly.